system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -6,29 +6,75 @@ const TopSitesEdit = require("./TopSitesEdit");
 const {TopSite, TopSitePlaceholder} = require("./TopSite");
 const CollapsibleSection = require("content-src/components/CollapsibleSection/CollapsibleSection");
 const ComponentPerfTimer = require("content-src/components/ComponentPerfTimer/ComponentPerfTimer");
+const {actionCreators: ac} = require("common/Actions.jsm");
 
-const TopSites = props => {
-  const realTopSites = props.TopSites.rows.slice(0, props.TopSitesCount);
-  const placeholderCount = props.TopSitesCount - realTopSites.length;
-  const infoOption = {
-    header: {id: "settings_pane_topsites_header"},
-    body: {id: "settings_pane_topsites_body"}
+/**
+ * Reducer that iterates through TopSites and counts types of images.
+ * @param acc Accumulator for reducer.
+ * @param topsite Entry in TopSites.
+ */
+function countTopSitesIconsTypes(topSites) {
+  const countTopSitesTypes = (acc, topsite) => {
+    const {
+      screenshot, showSmallFavicon,
+      smallFaviconFallback, tippyTopIcon
+    } = topsite;
+
+    if (screenshot && smallFaviconFallback) {
+      acc.screenshot++;
+    } else if (screenshot && showSmallFavicon) {
+      acc.screenshot_with_icon++;
+    } else if (tippyTopIcon) {
+      acc.tippytop++;
+    } else {
+      acc.rich_icon++;
+    }
+
+    return acc;
   };
-  return (<ComponentPerfTimer id="topsites" initialized={props.TopSites.initialized} dispatch={props.dispatch}>
-    <CollapsibleSection className="top-sites" icon="topsites" title={<FormattedMessage id="header_top_sites" />} infoOption={infoOption} prefName="collapseTopSites" Prefs={props.Prefs} dispatch={props.dispatch}>
-      <ul className="top-sites-list">
-        {realTopSites.map((link, index) => link && <TopSite
-          key={link.guid || link.url}
-          dispatch={props.dispatch}
-          link={link}
-          index={index}
-          intl={props.intl} />)}
-        {placeholderCount > 0 && [...Array(placeholderCount)].map((_, i) => <TopSitePlaceholder key={i} />)}
-      </ul>
-      <TopSitesEdit {...props} />
-    </CollapsibleSection>
-  </ComponentPerfTimer>);
-};
+
+  return topSites.reduce(countTopSitesTypes, {
+    "screenshot_with_icon": 0,
+    "screenshot": 0,
+    "tippytop": 0,
+    "rich_icon": 0
+  });
+}
+
+class TopSites extends React.PureComponent {
+  componentDidUpdate(props) {
```

Comment:
Can you also run this on `componentDidMount`? This won't cover the first load case